platform/mds/mds/src/main/java/org/motechproject/mds/osgi/MdsBundleWatcher.java

```diff
@@ -155,10 +155,34 @@ public void bundleChanged(BundleEvent event) {
                 bundlesToRefresh.add(bundle);
             }
         }
+        for(Bundle bundle : bundleContext.getBundles()) {
+            processForExtensions(outputs, bundle, schemaHolder);
+        }
 
         return outputs;
     }
 
+    private void processForExtensions(List<MDSProcessorOutput> outputs, Bundle bundle, SchemaHolder schemaHolder) {
+        if (skipBundle(bundle)) {
+            return;
+        }
+
+        synchronized (lock) {
+            // Before we process annotations, we wait until bundle resolves its dependencies
+            int count = 0;
```

Comment:
I belive this wait can be moved to a separate, private method to make the code cleaner.